### PR TITLE
(CDAP-7117) Added support for Spark Streaming Checkpoint

### PR DESCRIPTION
**Pull request by chtyim**
- Provides a Serializable SparkExecutionContext for the user Spark program

### CODE REVIEW

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkClassLoader.java [hunk 1]
```diff
@@ -102,14 +84,30 @@ public SparkRuntimeContext getRuntimeContext() {
...
-      throw new IllegalStateException("Creation of SparkExecutionContext is not allowed in the current context.");
+  public synchronized SparkExecutionContext getSparkExecutionContext(boolean createIfNotExists) {
+    if (sparkExecutionContext != null) {
+      return sparkExecutionContext;
     }
-    return contextFactory.create(runtimeContext);
+
+    if (!createIfNotExists) {
+      throw new IllegalStateException(
+        "SparkExecutionContext does not exists. " +
```
Comment (anwar6953) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkClassLoader.java]: nit - exist

File: cdap-unit-test/src/test/java/co/cask/cdap/spark/SparkStreamingTestRun.java [hunk 2]
```diff
@@ -45,16 +49,27 @@
   @ClassRule
   public static final KafkaTester KAFKA_TESTER = new KafkaTester();
 
+  @ClassRule
+  public static final TemporaryFolder TEMP_FOLDER = new TemporaryFolder();
+
+  @ClassRule
+  public static final TestConfiguration CONFIG = new TestConfiguration("explore.enabled", false);
```
Comment (anwar6953) [inline cdap-unit-test/src/test/java/co/cask/cdap/spark/SparkStreamingTestRun.java]: seems cleaner to use the Constants.Explore.EXPLORE_ENABLED

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkClassLoader.java (hunk 1 above)
Comment (chtyim) [reply cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkClassLoader.java]: fixed.

File: cdap-unit-test/src/test/java/co/cask/cdap/spark/SparkStreamingTestRun.java (hunk 2 above)
Comment (chtyim) [reply cdap-unit-test/src/test/java/co/cask/cdap/spark/SparkStreamingTestRun.java]: Changed. BTW, that constant should be called `Constants.Explore.ENABLED`.

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-DUT4718-3

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java [hunk 3]
```diff
@@ -172,17 +168,10 @@ protected void startUp() throws Exception {
       String metricsConfPath;
       String logbackJarName = null;
 
-      // Always copy (or link if local) user requested resources to local node. This is needed because the
-      // driver is running in the same container as the launcher (i.e. this process).
```
Comment (albertshau) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java:176]: is this not true anymore? don't understand how this changes where the driver is running. Or was this never true?

**Comment by albertshau** [comment]
just had one question, otherwise lgtm

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java (hunk 3 above)
Comment (chtyim) [reply cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java:176]: It is not true and not needed in distributed mode. The not true part is driver is no longer running in the same container as the launcher (client), and hence copying is not needed.

**Comment by albertshau** [comment]
lgtm

**Comment by chtyim** [comment]
Squashed and merging.

**Pull request merged**